server/graphql/v2/mutation/TransactionMutations.ts

```diff
@@ -23,6 +26,89 @@ const transactionMutations = {
       return legacyRefundTransaction(undefined, { id: transaction.id }, req);
     },
   },
+  rejectTransaction: {
+    type: new GraphQLNonNull(Transaction),
+    description: 'Rejects transaction, removes member from Collective, and sends a message to the contributor',
+    args: {
+      transaction: {
+        type: new GraphQLNonNull(TransactionReferenceInput),
+        description: 'Reference of the transaction to refund',
+      },
+      message: {
+        type: GraphQLString,
+        description: 'Message to send to the contributor whose contribution has been rejected',
+      },
+    },
+    async resolve(_, args, req): Promise<typeof Transaction> {
+      if (!req.remoteUser) {
+        throw new Unauthorized();
+      }
+
+      // get transaction info
+      const transaction = await fetchTransactionWithReference(args.transaction);
+
+      /** refund transaction and set status - - if the transaction has already been
+       * refunded we don't want to try and do it again, but we will continue with
+       * marking the order as 'REJECTED'
+       */
+      let refundedTransaction;
+      if (!transaction.RefundTransactionId) {
+        refundedTransaction = await legacyRefundTransaction(undefined, { id: transaction.id }, req);
+      } else {
+        refundedTransaction = await fetchTransactionWithReference({ legacyId: transaction.RefundTransactionId });
+      }
+      const orderToUpdate = await models.Order.update(
+        {
+          status: orderStatus.REJECTED,
+        },
+        {
+          where: { id: refundedTransaction.OrderId },
+          returning: true,
+        },
+      );
+      if (!orderToUpdate) {
+        throw new NotFound('Order not found');
+      }
+
+      // get membership info
+      const toAccount = await models.Collective.findByPk(transaction.CollectiveId);
+      const fromAccount = await models.Collective.findByPk(transaction.FromCollectiveId);
+      const memberships = await models.Member.findAll({
+        where: {
+          MemberCollectiveId: fromAccount.id,
+          CollectiveId: toAccount.id,
+          role: 'BACKER',
+        },
+      });
+
+      // remove member from Collective
+      if (memberships) {
+        memberships.forEach(membership => {
+          membership.destroy();
+        });
+      }
```

Comment:
Not a blocker, but did you try using a batch function as recommended in https://github.com/opencollective/opencollective-api/pull/4561#discussion_r489347538? I realized I made a mistake there (it should be `destroy`, not `delete`) but the idea is to do something tike:

```suggestion
      await models.Member.destroy({
        where: {
          MemberCollectiveId: fromAccount.id,
          CollectiveId: toAccount.id,
          role: 'BACKER',
        },
      });

      purgeCacheForCollective(fromAccount);
      purgeCacheForCollective(toAccount);
```

The main advantage of this would be to make only one SQL query, and to purge the cache only one time (vs doing it once per membership).

Reply:
Right, sorry, I didn't use that since I thought you needed to use `delete` with what you suggested. I can add this change.